### ISSUE TITLE
Weird response with range agg on float field

### ISSUE THREAD

**Issue by stratoula** [elastic/elasticsearch#81749]
## Description of the problem
We found a bug that is easy to reproduce it in kibana with Lens ranges on float field.
Specifically we created a float field and we apply the range aggregation on it. The request looks like that:

```
{
  "aggs": {
    "0": {
      "range": {
        "field": "test",
        "ranges": [
          {
            "to": 6,
            "from": 5
          },
          {
            "to": 10.6,
            "from": 6
          }
        ],
        "keyed": true
      }
    }
  },
  "size": 0,
  "fields": [],
  "script_fields": {},
  "stored_fields": [
    "*"
  ],
  "runtime_mappings": {},
  "_source": {
    "excludes": []
  },
  "query": {
    "bool": {
      "must": [],
      "filter": [],
      "should": [],
      "must_not": []
    }
  }
}
```
You can see that the requested ranges have decimal numbers.
The response we get though seems like that:
```
{
  "id": "FkUzMjU2dEt0VEJpeDNnTmNlSUgwSkEfRnBBb1BSWDdSTzY5U2szU3pPT1hDQToxMjAzNjgwMg==",
  "rawResponse": {
    "took": 0,
    "timed_out": false,
    "_shards": {
      "total": 1,
      "successful": 1,
      "skipped": 0,
      "failed": 0
    },
    "hits": {
      "total": 3,
      "max_score": null,
      "hits": []
    },
    "aggregations": {
      "0": {
        "buckets": {
          "5.0-6.0": {
            "from": 5,
            "to": 6,
            "doc_count": 1
          },
          "6.0-10.600000381469727": {
            "from": 6,
            "to": 10.600000381469727,
            "doc_count": 2
          }
        }
      }
    }
  },
  "isPartial": false,
  "isRunning": false,
  "total": 1,
  "loaded": 1,
  "isRestored": false
}
```
You can see that the second bucket is not the same with the response. 

**Elasticsearch version** (`bin/elasticsearch --version`):
7.16

**Steps to reproduce**:
 1.  Create an index and map the field to float
 ```
PUT myindex 
{
  "mappings": {
    "properties": {
      "test": {
        "type": "float"
      }
    }
  }  
}
```

```
PUT myindex/_bulk
{"index": {"_id": "1"}}
{"test": 10}
{"index": {"_id": "2"}}
{"test": 5.5}
{"index": {"_id": "3"}}
{"test": 6.7}
```

 2. Get the range agg with ranges that have decimal numbers
 3. Check the response

In 7.15.2 this worked fine. I think it was introduced in 7.16. (I replicated in a 7.16.1 instance)

**Comment by salvatore-campagna** [comment]
Maybe this is a consequence of the following https://github.com/elastic/elasticsearch/pull/78932?

**Comment by salvatore-campagna** [comment]
I tried to reproduce the issue with two unit tests...one using floats and one using doubles. It looks like that using doubles keys are "correct" ("5.0-6.0" and "6.0-10.6"), while they are not when using floats.

**Comment by salvatore-campagna** [comment]
All tests are green now. I will proceed with the backporting as soon as the review is completed. Will try to follow-up on this with the team so that maybe we can have a fix merged on Monday or Tuesday. I will backport this to: 7.16, 7.17 and 8.0.

**Comment by imotov** [comment]
Back porting to 7.17 and 8.0.0 is enough.

**Comment by prakx87** [comment]
We got the same issue in elasticsearch:7.16.2.

**Comment by knorpsiu** [comment]
I have this issue too. Strangely, if I use a range of `{"from": 0.0, "to": 0.5}` (and another `{"from": 0.5}`) it works...

**Comment by salvatore-campagna** [comment]
Back-porting to version `7.17.0` and `8.0.0` completed.

**Comment by salvatore-campagna** [comment]
Closing this as a result of merging the fix in https://github.com/elastic/elasticsearch/pull/81801 and back-porting the fix to version `7.17.0` and `8.0.0`.
